### PR TITLE
Virtual Component Groups in the Hierarchical Action List (2/2)

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

Put information about Virtual Component Groups in the Hierarchical Actions List.

https://www.pivotaltracker.com/story/show/181865548

### Important Notes

 - This PR implements the subtask 2 of 2 in the ["Virtual Component Groups in the Hierarchical Action List" task](https://www.pivotaltracker.com/story/show/181865548).
 - Note: the PR description does not include a screencast due to the changes in this PR not having any visual effect at this moment. The result of this PR's changes would be only observed in the Component Browser, but the Component Browser is not merged yet. As described in the task's Acceptance Criteria, the changes in this PR are currently only covered by tests.
 - Manual integration testing with the Engine showed that a response to an `executionContext/getComponentGroups` request is non-empty only after an `executionContext/executionComplete` message is received by the IDE. (See also [a discussion on Discord](https://discord.com/channels/401396655599124480/983669600854106112).) This was not known by the IDE team or documented before, and the existing code was modified to take this into account. The protocol docs are expected to be updated by the Engine team.
 - A list of component groups looked up in the suggestion database is cached in the node searcher as an optimization. 

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

**Comment by wdanilo** [comment]
Mateusz, can you include the screencast in the PR description or add an information why screencast is not needed, please?

**Comment by akavel** [comment]
@wdanilo done! (I added an explanation in the "Important Notes" section.)

File: app/gui/src/controller/searcher.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher.rs]: Add info where do they come from.

File: app/gui/src/controller/searcher.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher.rs]: There is a method "module_qualified_name" in mock::data, perhaps we can use it here?

File: app/gui/src/controller/searcher/component/group.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/group.rs]: I think the name like `from_execution_context_component_group` or `from_library_defined_component_group` is better, because it says "how to use this function" instead "how the function is implemented".

File: app/gui/src/controller/searcher/component/group.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/group.rs]: There is `Component::new` constructor for that.

File: app/gui/src/controller/searcher/component/builder.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/builder.rs]: There is some inconsistency in the builder's API - in one place we take suggestion entry ids and use our database to lookup entries, but here we take a ready group list that someone should prepare. I would move the responsibility of looking up things in suggestion db to the builder for each case.

Also, I would make the building favorites a separate method, instead of the argument of build (so the tests not interested in favorite section could skip this part).

File: app/gui/src/controller/searcher/component/builder.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/builder.rs]: The components from favorites section shall be also put in "all components" list, because we iterate over it when updating filtering. Also, the groups "update filtering" methods should be also called (I don't see a code for that).

File: app/gui/src/model/execution_context/plain.rs
Comment (farmaazon) [inline app/gui/src/model/execution_context/plain.rs:64]: I'm not sure where to do this, but It would be good to write info _when_ the component groups will be loaded. Perhaps it should be in the getter in API trait (so one would not be surprised, that component groups of newly created synchronized ExecutionContext are empty).

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/searcher/component/builder.rs
Comment (akavel) [reply app/gui/src/controller/searcher/component/builder.rs]: Ouch, I completely forgot about filtering 🤦  - thanks for catching this! ❤️ 

Please note that the favorites list will be filtered, but not sorted. I believe keeping the order of components in a favorites list is reasonable from a UX point of view, reinforcing a user's memory of the order of components in the virtual group as designed by the library author. From the implementation point of view, this is notably easier to do, as the library author's order of components would have to be restored from some extra storage after in-place sorting/filtering is disabled.

File: app/gui/src/controller/graph/executed.rs
Comment (wdanilo) [inline app/gui/src/controller/graph/executed.rs:182]: Why we don't either:
1. Users accessing `execution_ctx`?
2. Deref to `execution_ctx`?

File: app/gui/src/controller/searcher.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: I'm not sure I'm following this comment. The favorites are already set to something here, but where in the code they are not set yet? If we refer here as "already", it means that there is some code where they are not yet. Maybe we should link this code? Without this information, I feel this comment is incomplete.

File: app/gui/src/controller/searcher.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: 1. Please remember about splitting expressions in such a way that they are one liners if possible.
2. Do we need to. clone ref the database? It seems strange to me that this struct has two fields and both fields contain reference to the same thing. Maybe its needed, but still feels like a strange desing to me.

File: app/gui/src/controller/searcher.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: I don't think we should pass `Rc<Vec<...>>` as argument anywhere. This is an implementation detial that we keep this `Vec` in an Rc. Either create a type for it or at least a type alias. This is just `ImVec`.

File: app/gui/src/controller/searcher.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: The `set_favorites` function should be easier to use. You should be able to write `builder.set_favorites(groups))` instead.

File: app/gui/src/controller/searcher.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: What does this comment mean? How `reload_list` is used to "verify the contents"?

File: app/gui/src/controller/searcher/component.rs
Comment (wdanilo) [inline app/gui/src/controller/searcher/component.rs]: Please use Rust doc links to `ComponentBrowser` implementation, so when reading these docs we can click there to learn more aobut hwat component browser is.

**Review by wdanilo — CHANGES_REQUESTED** [review]
Please remember to apply these comments to all files and all places where they are applicable.

File: app/gui/src/controller/graph/executed.rs
Comment (farmaazon) [reply app/gui/src/controller/graph/executed.rs:182]: The graph::executed controller keeps synchronized execution context and graph, so the graph will always contain the definition of the top item on the execution context stack. So we don't want to give it to the user, which could desynchronize call stack with graph.

File: app/gui/src/controller/searcher.rs
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs]: I think the "already" meaning here is "you don't need to do it anymore" rather than "there is code where they're not set". Perhaps the term "prepopulated" will suit here.

I would also add an info _why_ we keep such builder.

File: app/gui/src/controller/searcher.rs
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs]: You can call `builder.set_favorites(&*groups))` but is it _easier_ is a matter of taste.

On the other hand, having the argument like `impl AsRef<impl IntoIterator<Item = ...>>` seems like an overkill to me.

File: app/gui/src/controller/searcher/component.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component.rs]: The `all_groups` should return all groups - including favorites. If we don't want to, the method should be renamed.

File: app/gui/src/controller/searcher/component.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component.rs]: Also, we check the matched entries' ordering - add info here.

File: app/gui/src/controller/searcher/component/builder.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/builder.rs]: This can be replaced by 
```rust
self.all_components.extend(&*group.entries.borrow());
``` 
Or something like that.

File: app/gui/src/controller/searcher/component/builder.rs
Comment (farmaazon) [inline app/gui/src/controller/searcher/component/builder.rs:155]: You may add a comment explaining, why favorites does _not_ need to be updated.

**Review by farmaazon — APPROVED** [review]

File: app/gui/src/controller/searcher/component.rs
Comment (akavel) [reply app/gui/src/controller/searcher/component.rs]: Actually, it appears that the name `all_groups` was not correct even before: it did not include the groups from the `top_modules` field either.

Given this, and that `all_groups` was only used in one place, I decided to inline its body. I find the resulting code more clear and explicit as to what it is actually doing.

File: app/gui/src/controller/searcher/component/builder.rs
Comment (akavel) [reply app/gui/src/controller/searcher/component/builder.rs:155]: As we agreed on Discord, it seems enough that the method's docs already mention that favorites will *not* be updated by the method:
```
    /// ... (Does not sort
    /// the [`component::List::favorites`].)
```

File: app/gui/src/controller/searcher/component.rs
Comment (farmaazon) [reply app/gui/src/controller/searcher/component.rs]: > Actually, it appears that the name `all_groups` was not correct even before: it did not include the groups from the `top_modules` field either.

That's not true. There was no need of iterating over top_modules, as they repeated the groups contained in the map. Only flattened submodules have no representation in that map (perhaps it should be commented somewhere).

Comment (akavel) [reply app/gui/src/controller/searcher/component.rs]: > > Actually, it appears that the name `all_groups` was not correct even before: it did not include the groups from the `top_modules` field either.
> 
> That's not true. There was no need of iterating over top_modules (...)

Understood, thanks! I reverted the inlining and renamed the method.

**Review by wdanilo — APPROVED** [review]

**Comment by wdanilo** [comment]
I did not check the logic. If @farmaazon is OK with it, we can merge it :)

**Comment by akavel** [comment]
(*Process note:* the code-review is now Approved, therefore the PR is waiting for an Acceptance/QA Review by @farmaazon  on Task https://www.pivotaltracker.com/story/show/181865548 before merge.)

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/181865548">Virtual Component Groups in the Hierarchical Action List - Pivotal Tracker</a></strong></div></blockquote>

**Pull request merged**
